hi-ml-cpath/src/health_cpath/utils/girder.py

```diff
@@ -467,21 +521,48 @@ def upload(
         search_mode = "text" if search_mode == "full" else search_mode
         progress = tqdm(unique_slide_ids)
         responses = []
-        for slide_id in progress:
-            progress.set_description(slide_id)
-            if id_filter not in slide_id:
-                continue
-            item = dsa.search_item(slide_id, search_mode=search_mode)
-            tqdm.write(f"Processing slide {slide_id} - {item.url}")
-            annotation_name = self.run.id
-            annotation = self.get_slide_annotation_from_df(
-                self.df,
-                slide_id,
-                annotation_name,
-                **annotation_kwargs,
-            )
-            if not dry_run:
-                responses.append(item.add_annotation(annotation))
+        if folder:
+            folder_id = dsa.get_folder_id(folder)
+            items = dsa.get_items_in_folder(folder_id)
+            print(f"Found a total of {len(items)} items in folder {folder}")
+            for slide_id in progress:
+                progress.set_description(slide_id)
+                if id_filter not in slide_id:
+                    continue
+                matching_items = [item for item in items if slide_id in item.name]
+                if not matching_items:
+                    print(f"No items in DSA for slide ID {slide_id}")
+                    continue
+                elif len(matching_items) > 1:
+                    print(f"Multiple items in DSA for slide ID {slide_id}. Skipping this slide.")
+                    continue
+                item = matching_items[0]
+                tqdm.write(f"Processing slide {slide_id} - {item.url}")
+                annotation_name = f"{self.run.id} ({self.run.display_name})"
+                annotation = self.get_slide_annotation_from_df(
+                    self.df,
+                    slide_id,
+                    annotation_name,
+                    **annotation_kwargs,
+                )
+                if not dry_run:
+                    responses.append(item.add_annotation(annotation))
+        else:
+            for slide_id in progress:
+                progress.set_description(slide_id)
+                if id_filter not in slide_id:
+                    continue
+                item = dsa.search_item(slide_id, search_mode=search_mode)
+                tqdm.write(f"Processing slide {slide_id} - {item.url}")
+                annotation_name = self.run.id
+                annotation = self.get_slide_annotation_from_df(
+                    self.df,
+                    slide_id,
+                    annotation_name,
+                    **annotation_kwargs,
+                )
+                if not dry_run:
+                    responses.append(item.add_annotation(annotation))
```

Comment:
line 551 - 565 and 525 - 549 have a lot in common we can write a function for that, that takes item as an argument and then do the item selection separately in case a folder is specified line 532 - 539

Reply:
Good point - I've refactored to put the loop first, which also simplifies a lot.